documentation/adoc/cluster-controller.adoc

```diff
@@ -241,6 +266,97 @@ configuration file. For this reason, you can find more information on how to use
+Optional with no default.
+`limits/memory`::
+the memory request for the container, corresponding directly to Kubernetes' {resources-link}[`spec.containers[\].resources.limits.memory`] setting.
+Optional with no default.
+`limits/cpu`::
+the cpu request for the container, corresponding directly to Kubernetes' {resources-link}[`spec.containers[\].resources.limits.cpu`] setting.
+Optional with no default.
+
+[[jvm_json_config]]
+===== JVM Options
+
+It is possible to configure a subset of available JVM options on Kafka, Zookeeper and Topic Controller containers using a JSON object.
+The object has a property for each JVM (`java`) option which can be configured:
+
+`-Xmx`::
+The maximum heap size. See the <<setting_xmx>> section for further details.
+
+`-Xms`::
+The initial heap size.
+Setting the same value for initial and maximum (`-Xmx`) heap sizes avoids the JVM having to allocate memory after startup, at the cost of possibly allocating more heap than is really needed.
+For Kafka and Zookeeper pods such allocation could cause unwanted latency.
+For Kafka Connect avoiding over allocation may be the more important concern, especially in distributed mode where the effects of over-allocation will be multiplied by the number of consumers.
+
+NOTE: The units accepted by JVM settings such as `-Xmx` and `-Xms` are those accepted by the OpenJDK `java` binary in the corresponding image.
+Accordingly, `1g` or `1G` means 1,073,741,824 bytes.
+This is in contrast to the units used for <<resources_json_config,memory limits and requests>>, which follow the Kubernetes convention where `1G` means 1,000,000,000 bytes, and `1Gi` means 1,073,741,824 bytes
+
+.Example Resource limits and requests JSON configuration
+[source,json]
+----
+{
+  "-Xmx": "2g",
+  "-Xms": "2g"
+}
+----
+
+In the above example, the JVM will use 2 GiB (=2,147,483,648 bytes) for its heap.
+Its total memory usage will be approximately 8GiB.
+
+[[setting_xmx]]
+====== Setting `-Xmx`
+
+The default value used for `-Xmx` depends on whether there is a <<resources_json_config,memory request>> for the container:
+
+* If there is a memory request, the JVM's maximum memory will be limited according to the kind of pod (Kafka, Zookeeper, TopicController) to an appropriate value less than the request.
+* Otherwise, when there is no memory request, the JVM's maximum memory will be limited according to the kind of pod and the RAM available to the container.
+
+[IMPORTANT]
+====
+Setting `-Xmx` explicitly is requires some care:
+
+* The JVM's overall memory usage will be approximately 4 × the maximum heap as configured by `-Xmx`.
+
+* If `-Xmx` is set without also setting an appropriate Kubernetes
+memory request, it is possible that the container will be killed should the Kubernetes node
+experience memory pressure (from other Pods running on it).
+====
+
+When `-Xmx` is set explicitly, it is recommended use a memory request that is at least 4.5 × the request heap memory when setting `-Xmx`.
```

Comment:
Try running `docker run -m 1GB openjdk:8u162 java -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap -XX:MaxRAM=1g -XshowSettings:vm -version`

For me that shows:

```
VM settings:
    Max. Heap Size (Estimated): 228.00M
    Ergonomics Machine Class: server
    Using VM: OpenJDK 64-Bit Server VM

openjdk version "1.8.0_162"
OpenJDK Runtime Environment (build 1.8.0_162-8u162-b12-1~deb9u1-b12)
OpenJDK 64-Bit Server VM (build 25.162-b12, mixed mode)
```

So 1G max ram results in ~¼G max heap. The use of `-XX:+UseCGroupMemoryLimitForHeap` means it's not entirely linear wrt MaxRAM, the memory given to `docker` also makes a difference.

Reply:
This is very different from what we do in the controllers: https://github.com/strimzi/java-docker-base/blob/master/scripts/dynamic_resources.sh

There we seem to take the assumption that -Xmx is 50% of container memory.

Reply:
I agree, it *is* different. I do not claim to be an expert in configuring the JVM. A lot of this stuff isn't officially documented so you end up reading blogs by people who might not be experts either. 

Maybe we can find an expert.

Reply:
I'm not an expert either, but the settings I've used elsewhere is to set Xmx to 50% of the container memory limit. The JVM does tend to default to around 25% when the correct CGroup settings are enabled, but that seems to be too conservative for most cases.

Also, the heap should be set in relation to the container memory limit - not the container memory request. It's not fatal if a container exceeds its requested memory but it is fatal if a container tries to exceed its memory limit.

Reply:
Well, the request is guaranteed while the limit might not be available. That can cause issues if you configure the memory based on the limit, or? Especially with a stateful apps like Kafka and Zookeeper I would think that this is better to be avoided.

Reply:
You're right that request is guaranteed (or, if it can't be guaranteed, the pod won't be scheduled to a node until the request can be met). Since request is always lower than limit, you can set Xmx based on the request. You're just giving up the opportunity to potentially use more memory than what you requested. I've only ever seen containers OOMKilled by Kube/OpenShift when they exceed their memory limit. Exceeding the request isn't a big deal as the clusters are designed to overcommit memory to some extent.

Because it's complicated, I usually end up just setting request and limit to the same value and sizing my heap manually.